Initial implementation of KImg component

thanksameeelian opened the pull request:
## Description
This is the initial implementation of the KImg component, which will ultimately fulfill all the specifications of [the KImg epic](https://github.com/learningequality/kolibri-design-system/issues/368), but for this round only includes `src`, `altText`, `isDecorative`, `height`, `width`, `min-width`, `max-width`, `min-height`, and `max-height`.

#### Issue addressed

Addresses #369 


## Steps to test

1. `yarn dev`
2. `KImg` doc page should be accessible at http://localhost:4000/kimg
3. can edit any of the examples in docs/pages/kimg.vue to test functionality and see results reflected on the KImg doc page


## Testing checklist

- [X] Contributor has fully tested the PR manually


## Reviewer guidance
<!-- Delete anything that doesn't apply so your reviewer knows what to check for -->

- [ ] Is the code clean and well-commented?
- [ ] Are there tests for this change?
- [ ] Are all UI components LTR and RTL compliant (if applicable)?
- [ ] _Add other things to check for here_

## Comments
<!-- Any additional notes you'd like to add -->

MisRob commented:
@jtamiace if you'd like to read through the documentation, we'd appreciate your input

MisRob commented:
Documentation preview https://deploy-preview-404--kolibri-design-system.netlify.app/kimg/

akolson commented on docs/tableOfContents.js:
```diff
@@ -232,6 +232,11 @@ export default [
+      new Page({
+        path: '/kimg',
+        title: 'KImg',
+        isCode: true,
```
It would be nice to add the `keywords` property to allow for better searching within the documentation. For example, currently when you try to search an image component using the "image" keyword, the search would yield no results. Please see usages within this file for implementation details

akolson replied:
I think the changes made should suffice. Thanks @thanksameeelian. Non-blocking but would be good to follow existing patterns i.e. pass variable to keywords as opposed to the actual array object

thanksameeelian replied:
thanks! i checked out that pattern when i was adding the keywords here, but only saw the pattern of using a variable when multiple items had the same keywords, so the variable was being used in at least two places. 

i don't think that pattern is applicable here, as `KImg` is the only item using those specific keywords. i am following the pattern i saw for each item that had a unique list of variables, as in those cases the array itself was passed. you can see this patten in `KSelect` and `KSwitch` in the table of contents. i think this makes sense because it avoids unnecessary abstraction. 

but if another item were to share the same keywords at kimg, i would totally agree that a variable would be the right call!

akolson commented on docs/pages/kimg.vue:
```diff
@@ -0,0 +1,113 @@
+            :src="require('../assets/img_sample_for_kimg.png')"
+            altText="Computers are run by bees"
+            :height="'250px'"
+            :width="225"
```
It would be good to add the entire snippet(including the `KImg` tag as well) and not just the properties for better readability and usability from the developers' perspective. Also, wrapping `<DocsShowCode>{Your snippet}</DocShowCode>` around the snippet would give it a better formatting.

akolson commented on docs/pages/kimg.vue:
```diff
@@ -0,0 +1,113 @@
+            :src="require('../assets/img_sample_for_kimg.png')"
+            isDecorative
+            :height="50"
+            :width="50"
```
Please see the comment above

akolson commented on lib/KImg.vue:
```diff
@@ -0,0 +1,159 @@
+
+  <div>
+    <img
+      id="kimg"
```
Using id could prove problematic especially when more than one `KImg` is used on a single page. It is on this basis that css class referencing is recommended instead

akolson commented on lib/KImg.vue:
```diff
@@ -0,0 +1,159 @@
+          '--min-height': this.imgMinHeight,
+          '--max-width': this.imgMaxWidth,
+          '--min-width': this.imgMinWidth,
+        };
```
This seems like neat way to achieve the styling, however it looks like [css variables are not supported in IE 11](https://caniuse.com/css-variables) and this would definitely break compatibility.  [Class & style binding](https://vuejs.org/guide/essentials/class-and-style.html) might be a better approach to this

akolson commented on lib/KImg.vue:
```diff
@@ -0,0 +1,159 @@
+            return `${propValue}px`;
+          } else {
+            // split numbers apart from units
+            const [, ...arr] = propValue.match(/(\d*)([\s\S]*)/);
```
@thanksameeelian Awesome approach 💯  :) This was so much fun and exciting to review. It felt like a sprinkling of magic dust!

akolson commented on lib/KImg.vue:
```diff
@@ -0,0 +1,159 @@
+    min-height: var(--min-height);
+    max-width: var(--max-width);
+    min-width: var(--min-width);
+  }
```
Please see `cssVars` comment

akolson commented on lib/KImg.vue:
```diff
@@ -0,0 +1,159 @@
+    methods: {
+      validateAndFormatUnits(propValue) {
+        if (propValue) {
+          if (Number.isInteger(propValue)) {
```
This check would fail for a `propValue` that is a floating point number(eg 30.5).  The same `propValue` but with a unit(eg 30.5rem) would pass the validation. This not only causes an inconsistency that needs to be streamlined but also means that the regex fails for floating point numbers with units(eg 30.5rem). Maybe using the `isNaN` is a better option. This would also mean updating the regex to capture floating point numbers as well. I tweaked your regex to cater for the decimal points part and the final result is this `(\d*\.\d+)([\s\S]*)`

akolson replied:
This is another update to my earlier recommended regex. using `(\d*\.?\d+)([a-zA-Z,\s]*)`  instead will ensure that values like 30px, 12.5rem would pass validation. However values like 10.em would fail validation as such an input would not work when applied in css.

akolson commented:
Hi @thanksameeelian I reviewed your code alongside @MisRob and we think your code generally looks good! Great Job :). We found few issues for you to look into. Also, to give us better design perspective especially on the documentation, we've added @jtamiace to give her thoughts on  https://deploy-preview-404--kolibri-design-system.netlify.app/

akolson commented:
Something that we pondered about was whether it was necessary to widen the scope of units used. At the moment, only `px`, `em` `rem`, `vh` are being used. @marcellamaki We are happy to hear your thoughts

akolson reviewed (changes requested):
Please see comments

thanksameeelian commented:
> Hi @thanksameeelian I reviewed your code alongside @MisRob and we think your code generally looks good! Great Job :). We found few issues for you to look into. Also, to give us better design perspective especially on the documentation, we've added @jtamiace to give her thoughts on https://deploy-preview-404--kolibri-design-system.netlify.app/kimg/

thanks so much @akolson & @MisRob! i've read through your comments & i don't think i have any questions about them, but will be sure to ask if i end up needing clarification. i'll get on these changes but also keep my eyes open for feedback from the design team!

thanksameeelian commented:
> Something that we pondered about was whether it was necessary to widen the scope of units used. At the moment, only `px`, `em` `rem`, `vh` are being used. @marcellamaki We are happy to hear your thoughts

it was just an oversight that i didn't include a more exhaustive list of units initially, so i'm def happy to update and try to catch 'em all! 😀

marcellamaki commented:
> Something that we pondered about was whether it was necessary to widen the scope of units used. At the moment, only `px`, `em` `rem`, `vh` are being used. @marcellamaki We are happy to hear your thoughts

@akolson @MisRob @thanksameeelian - I think having more units would be great idea! At the same time, in my opinion it isn't necessarily a **blocker**, unless we can think of a specific use case in one of our products right now where we would likely be needing/using some other unit. If so, that should be prioritized - otherwise I see no problem with expanding this on an as-needed basis.

Thanks Amelia for this great work, and Misha and Samson for the very thoughtful and comprehensive review! Very exciting to see this happening :)

jtamiace commented:
This looks great!  In terms of additional content, I'm not 100% sure if this suggestion is appropriate for this page which is focused on technical guidance and documentation, but I'm wondering if it makes sense to record any general patterns we use for missing images or error states if an image is broken, or if this is something that's out of the scope of KImg and handled elsewhere. For example in Studio and the Learning Platform we use a light background gray along with a centered icon of the resource type or a generic "image" icon.

MisRob commented:
One unit I'd suggest adding would be `%` since it's used commonly

thanksameeelian commented:
> This looks great! In terms of additional content, I'm not 100% sure if this suggestion is appropriate for this page which is focused on technical guidance and documentation, but I'm wondering if it makes sense to record any general patterns we use for missing images or error states if an image is broken, or if this is something that's out of the scope of KImg and handled elsewhere. For example in Studio and the Learning Platform we use a light background gray along with a centered icon of the resource type or a generic "image" icon.

thank you so much for your feedback jessica!! we've split this project up over several prs ([here's the epic!](https://github.com/learningequality/kolibri-design-system/issues/368)) so creating a placeholder image will be handled in the next pr, and additional error states potentially after that. i'd really appreciate your feedback on those once they're ready if you have the time, so i'll be sure to add examples of them in the doc & tag you/request your review!

thanksameeelian commented:
thanks so much for the initial round of review, @akolson, @MisRob, @marcellamaki, & @jtamiace!! i made all the changes indicated and would love to hear your thoughts on this new version! in terms of CSS units, i added `%` (and adjusted the regex) as well as all of the units i think i've encountered in the wild, based on [lists provided here](https://developer.mozilla.org/en-US/docs/Learn/CSS/Building_blocks/Values_and_units) - i'd be happy to list all of them if you think that's appropriate, but just didn't want to create anything overwhelming at the moment 🙂 really interested to hear your thoughts!

akolson reviewed (approved):
Hi @thanksameeelian,  Thanks for the changes, they LGTM! @MisRob @jtamiace, any other thoughts? Also @marcellamaki , at what point do we do manual QA on this? Should we wait for the component to develop further or have QA take a first look?

marcellamaki commented:
While I think some manual QA would be ideal in most situations, since we don't an immediate github issue in Kolibri that requires this as a fix, and we don't have plans to refactor immediately, I think it's okay to proceed without manual QA for now. There have been a lot of eyes on this, and thoughtful work and feedback, so I feel okay about moving forward without that testing. And, we should remain open to the possibility that some tweaks may be needed as we start integrating this into our products, and that's okay 😄

MisRob reviewed (approved):
Thank you @thanksameeelian, good work.

I too feel fine about doing manual QA as soon as the component is complete, as some upcoming issues may perhaps require tweaking code here and there a bit in areas that were implemented before.
